protocols/tls13/src/cryptolib.rs

```diff
@@ -0,0 +1,271 @@
+    match group_name {
+        NamedGroup::SECP256r1 => {
+            let Affine(X, Y) = p256_secret_to_public(P256Scalar::from_byte_seq_be(x));
+            Ok(X.to_byte_seq_be().concat(&Y.to_byte_seq_be()))
+        }
+        NamedGroup::X25519 => Ok(DHPK::from_seq(&x25519_secret_to_public(
+            SerializedScalar::from_seq(x),
+        ))),
+    }
+}
+
+pub fn ecdh(group_name: &NamedGroup, x: &DHSK, y: &DHPK) -> Res<KEY> {
+    match group_name {
+        NamedGroup::SECP256r1 => {
+            let pk = Affine(
+                P256FieldElement::from_byte_seq_be(y.slice_range(0..32)),
+                P256FieldElement::from_byte_seq_be(y.slice_range(32..64)),
+            );
+            let Affine(X, Y) = p256_point_mul(P256Scalar::from_byte_seq_be(x), pk);
+            Ok(X.to_byte_seq_be().concat(&Y.to_byte_seq_be()))
+        }
+        NamedGroup::X25519 => Ok(DHPK::from_seq(&x25519_point_mul(
+            SerializedScalar::from_seq(x),
+            SerializedPoint::from_seq(y),
+        ))),
+    }
+}
+
+pub fn hash(ha: &HashAlgorithm, payload: &Bytes) -> Res<HASH> {
+    match ha {
+        HashAlgorithm::SHA256 => Ok(HASH::from_seq(&sha256(payload))),
+        HashAlgorithm::SHA384 => Err(unsupported_algorithm),
+    }
+}
+
+pub fn hmac(ha: &HashAlgorithm, mk: &MACK, payload: &Bytes) -> Res<HMAC> {
+    match ha {
+        HashAlgorithm::SHA256 => Ok(HMAC::from_seq(&hmac::hmac(mk, payload))),
+        HashAlgorithm::SHA384 => Err(unsupported_algorithm),
+    }
+}
+
+pub fn hmac_verify(ha: &HashAlgorithm, mk: &MACK, payload: &Bytes, m: &HMAC) -> Res<()> {
+    // XXX: this is pretty ugly
+    let my_hmac = hmac(ha, mk, payload)?;
+    let mut result = if m.len() == my_hmac.len() {
+        Ok(())
+    } else {
+        Err(mac_failed)
+    };
+    for i in 0..m.len() {
+        if !my_hmac[i].equal(m[i]) {
+            result = Err(mac_failed);
+        }
+    }
+    result
+}
+
+pub fn verk_from_cert(cert:&Bytes) -> Res<VERK> {
+    Ok(VERK::new(64))
+}
+
+pub fn sign(sa: &SignatureScheme, ps: &SIGK, payload: &Bytes) -> Res<SIG> {
+    return Ok(SIG::new(32));
+}
+pub fn verify(sa: &SignatureScheme, pk: &VERK, payload: &Bytes, sig: &Bytes) -> Res<()> {
+    return Ok(());
+}
+
+pub fn hkdf_extract(ha: &HashAlgorithm, k: &KEY, salt: &KEY) -> Res<KEY> {
+    match ha {
+        HashAlgorithm::SHA256 => Ok(KEY::from_seq(&hkdf::extract(salt, k))),
+        HashAlgorithm::SHA384 => Err(unsupported_algorithm),
+    }
+}
+
+pub fn hkdf_expand(ha: &HashAlgorithm, k: &KEY, info: &Bytes, len: usize) -> Res<KEY> {
+    match ha {
+        HashAlgorithm::SHA256 => Ok(KEY::from_seq(&hkdf::expand(k, info, len))),
+        HashAlgorithm::SHA384 => Err(unsupported_algorithm),
+    }
+}
+
+pub fn aead_encrypt(
+    a: &AEADAlgorithm,
+    k: &AEK,
+    iv: &AEIV,
+    payload: &Bytes,
+    ad: &Bytes,
+) -> Res<Bytes> {
+    // XXX: the result should be Seq<u8> not Seq<U8>.
```

Comment:
Why so ? Since the result is encrypted, it should be comprised of secret bytes and not public bytes no?

Reply:
Its a choice on whether we declassify ciphertexts or not, i think we can go either way, but in the TLS code I assume that *all* arrays are Seq<U8>

Reply:
`Seq` are fine, the choice would be between `Seq<u8>` and `Seq<U8>`. I would prefer the latter since it enforces secret integers discipline.

Reply:
Right, so there's currently a mix of public and secret versions for ciphertexts and tags. Because we want to compare tags for example we might want them to be public. Also from a crypto perspective these are public values, not secret values.